src/mbgl/gl/shader.cpp

```diff
@@ -21,20 +21,24 @@ Shader::Shader(const char* name_,
                const char* vertexSource,
                const char* fragmentSource,
                Context& context,
-               Defines defines)
+               gl::ShaderParameters& shaderParameters)
     : name(name_),
       program(context.createProgram()),
       vertexShader(context.createVertexShader()),
       fragmentShader(context.createFragmentShader()) {
     util::stopwatch stopwatch("shader compilation", Event::Shader);
 
-    if (!compileShader(vertexShader, vertexSource)) {
+    std::string vertex(vertexSource);
+    vertex.replace(vertex.find_first_of('\n'), 1,"\n#define DEVICE_PIXEL_RATIO " + std::to_string(shaderParameters.pixelRatio) + "\n");
```

Comment:
I'll take that back; we are initializing `Map` with a constant pixel ratio that can't be changed after the fact anyway.